Custom DST (#2)
kennytm/rfcs: text/0000-dyn-type.md

            - [Module `core::dst`](#module-coredst)
            - [Trait `dst::Regular`](#trait-dstregular)
            - [Trait `dst::Inline`](#trait-dstinline)
            - [Trait `dst::Derived`](#trait-dstderived)
            - [Function `reduce`](#function-reduce)
        - [Alternative syntax](#alternative-syntax)
- [Prior art](#prior-art)
    - [Previous RFCs](#previous-rfcs)
    - [C and C++](#c-and-c)
- [Unresolved questions](#unresolved-questions)
    - [Safe construction](#safe-construction)
- [Future possibilities](#future-possibilities)
    - [`DynSized`](#dynsized)
    - [`#[repr]` as a constraint](#repr-as-a-constraint)
        - [Aligned foreign type](#aligned-foreign-type)
    - [Copy initialization](#copy-initialization)

<!-- /TOC -->

</details>

# Motivation
[motivation]: #motivation

As of Rust 1.31, we only have 4 families of DSTs:
slices (`[T]` and `str`), trait objects (`dyn Trait`), extern types and structs/tuples with a DST tail.
However this is not enough, e.g. we cannot easily define a length-prefix array or a 2D matrix slice with this.
And recently this even leads to attempt trying to stabilize extern types with questionable semantics

> (e.g. fixing the alignment to be 1 even if it is completely opaque) in order to make custom DSTs possible.

Reviewer: ```suggestion
(e.g. fixing the alignment to be 1 even if it is completely opaque) to make custom DSTs possible.
```

This RFC is submitted to providing a proper way to define a true custom DST,
such that the internals can be safely defined.

See the [What do we want to support?](#what-do-we-want-to-support) section for more details.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## Tutorial

> Even if this is a “guide-level” explanation, readers are still expected to understand what is a dynamic-sized type
> and why slices `&[T]` and trait objects `&dyn Trait` are implemented as [fat pointers].
>
> These can be referred in TRPL 2/e [§4.3], [§17.2], and also Rustonomicon [§2.2].

[§4.3]: https://doc.rust-lang.org/book/second-edition/ch04-03-slices.html
[§17.2]: https://doc.rust-lang.org/book/second-edition/ch17-02-trait-objects.html
[§2.2]: https://doc.rust-lang.org/nomicon/exotic-sizes.html
[fat pointers]: https://www.google.com/search?tbm=isch&q=fat+pointer

### Dynamic sized types

Dynamic-sized types (DSTs) are types where the memory structure cannot be sufficiently determined at compile time.

> Memory structure means the allocation size, data alignment, destructor, etc. In order to know these,

Reviewer: ```suggestion
Memory structure means the allocation size, data alignment, destructor, etc. To know these,
```

pointers to DSTs will need to carry additional *metadata* to supplement the calculation at runtime.
In Rust there are two built-in DSTs:

* Slices `[T]` and string slice `str` — you need to know how long the slice is, so the metadata is the slice length.
* Trait objects `dyn Trait` — you need to know the concrete type, so the metadata is information about the type itself
    (size, alignment, destructor function pointer, and trait vtable).

Introducing DSTs in Rust allows us to separate the concern of “where to place these memory as a whole” vs
“how to use the content in this memory region”.

> This allows, e.g. slices be easily supported by any resource management schemes, be it a raw pointer `*[T]`,

Reviewer: ```suggestion
This allows, e.g. slices to be easily supported by any resource management schemes, be it a raw pointer `*const [T]` or `*mut [T]`,
```

shared reference `&[T]`, owned box `Box<[T]>`, ref-counted container `Rc<[T]>`,

> even third-party smart pointers like `Gc<'root, [T]>`.

Reviewer: ```suggestion
and even third-party smart pointers like `Gc<'root, [T]>`.
```

### The case for custom DSTs

Slices and trait objects are not flexible enough for every use case. Sometimes you need to define your own *custom DST*.
An example is a 2D matrix. If we do not have custom DSTs, we would represent a rectangular sub-matrix as

```rust ,ignore
struct MatrixRef<'a, T> {
    elements: &'a [T],
    width: usize,
    height: usize,
    stride: usize,
}
```

This structure is not scalable to additional smart pointers —
you’ll need a `BoxMatrix<T>` to support elements stored in a `Box<[T]>`, a `RcMatrix<T>` for elements in `Rc<[T]>`, etc.
One may attempt to turn this type into a DST itself by storing elements as a flexible array member
at the end of the structure:

```rust ,ignore
struct MatrixData<T> {
    width: usize,
    height: usize,
    stride: usize,
    elements: [T],
}
```

> This, however, has a problem that cannot be sub-sliced,

Reviewer: What does it mean to sub-slice a problem? To make it into sub-problems?

Author: Rewording to

> However, this matrix cannot be sliced into smaller sub-matrices,

because `Index` requires you to return a reference without allocation!

```rust ,ignore

> impl<T> Index<Range<usize>, Range<usize>> for MatrixData<T> {

Reviewer: ```suggestion
impl<T> Index<(Range<usize>, Range<usize>)> for MatrixData<T> {
```

    type Output = Self;
    fn index(&self, index: (Range<usize>, Range<usize>)) -> &Self {
        // Ok, how could you obtain a sub-matrix of `self`
        // while placing the `width`, `height`, `stride` members
        // in the correct position?
        //
        // Try it. No, it is impossible without allocation.
        panic!()
    }
}
```

> The correct way is to make the width, height and stride metadata themselves, placed outside of the elements array:

Reviewer: ```suggestion
The correct way is to make the width, height, and stride metadata themselves, placed outside of the elements array:
```

Sponsored by Oxford commas. ;)

    then somehow convince the compiler to treat every `*Mat<T>` as `(*T, MatMeta)`
"}
```

### Creating custom DSTs

Starting from Rust 1.XX, we are allowed to define custom DSTs with arbitrary metadata:

```rust
dyn type Mat<T>(regular: MatMeta) = [T];
```

This tells the compiler that:

* The pointer itself will carry a metadata represented by the type `MatMeta`.
* The type is a *regular DST*, explained below.
* A pointer to a matrix `*Mat<T>` can be safely *reduced* to the corresponding built-in DST `*[T]`.
    The metadata `MatMeta` of `Mat<T>` can be converted into the corresponding metadata of `[T]`
    (`usize`, the length of the slice).

In memory it looks like this:

![](media/dyn-type/mat-ref-repr.svg)

We can calculate the actual size of the matrix using the width, height and stride alone:

```rust ,ignore
impl MatMeta {
    const fn len(&self) -> usize {
        if self.height == 0 {
            0
        } else {
            (self.height - 1) * self.stride + self.width
        }
    }
}
```

and we call this kind of DSTs a *regular DST*. We provide the size by implementing the `std::dst::Regular` trait:

```rust ,ignore
impl<T> std::dst::Regular for Mat<T> {

>     const fn reduce_with_meta(meta: MatMeta) -> usize {

Reviewer: `const fn` isn't supported in traits yet...

Author: I'll move the mention about RFC #2237 further up then.

        meta.len()
    }
}
```

### Using the DST

> The metadata of our DST contains essential information how to retrieve the elements.

Reviewer: ```suggestion
The metadata of our DST contains essential information on how to retrieve the elements.
```

```rust ,ignore
use std::ptr;

impl<T> Mat<T> {
    pub fn width(&self) -> usize {
        ptr::metadata(self).width
    }
}
```

Accessing the content could be via by *reducing* the custom DST back into a built-in DST.

```rust ,ignore
use std::{ptr, dst};

impl<T> Mat<T> {
    pub fn iter<'a>(&'a self) -> impl Iterator<Item = &'a T> {
        let meta = ptr::metadata(self);
        let slice = dst::reduce(self); // convert `&Mat<T>` into `&[T]`
        slice.chunks(meta.stride).flat_map(|c| &c[..meta.width])
    }

    pub fn iter_mut<'a>(&'a mut self) -> impl Iterator<Item = &'a mut T> {
        let meta = ptr::metadata(self);
        let slice = dst::reduce_mut(self); // convert `&mut Mat<T>` into `&mut [T]`
        slice.chunks_mut(meta.stride).flat_map(|c| &mut c[..meta.width])
    }
}
```

### Unsizing

Built-in Rust DSTs like slice and trait objects support *unsize coercion*,
which allows you to turn a sized pointer into an unsized pointer:

```rust ,ignore
let array: &[u32; 4] = &[1, 2, 3, 4];
let slice: &[u32] = array;  // <-- coercion: `&[u32; 4]` becomes `&[u32]`!
```

This makes constructing a new DST much easier! Unsize coercion works by transforming the metadata

> value (here, from `()` to `4`), and reinterpret the pointed memory as the target type.

Reviewer: ```suggestion
value (here, from `()` to `4`), and reinterpreting the pointed memory as the target type.
```

We could write these rules for our custom DSTs as well, through the `std::marker::Unsize` trait:

```rust ,ignore
unsafe impl<T, const width: usize, const height: usize> Unsize<Mat<T>> for [[T; width]; height] {
    fn unsize(_: ()) -> MatMeta {
        MatMeta {
            width,
            height,
            stride: width,
        }
    }
}
```

> Unsizing works not just between sized and unsized type, it can be done between two DSTs too!

Reviewer: ```suggestion
Unsizing works not just between sized and unsized types, it can be done between two DSTs too!
```

        let meta = ptr::metadata(self);
        assert!(index.0 < meta.width, "column index out of bounds");
        assert!(index.1 < meta.height, "row index out of bounds");
        let index = meta.to_linear_index(index.0, index.1);
        &dst::reduce(self)[index]
    }
}
```

Now try to fetch a rectangular sub-matrix:

```rust ,ignore
impl<T> Index<(Range<usize>, Range<usize>)> for Mat<T> {
    type Output = Self;
    fn index(&self, index: (Range<usize>, Range<usize>)) -> &Self {
        let meta = ptr::metadata(self);
        assert!(index.0.start <= index.0.end, "invalid column range");
        assert!(index.0.end <= meta.width, "column index out of bounds");
        assert!(index.1.start <= index.1.end, "invalid row range");
        assert!(index.1.end <= meta.height, "row index out of bounds");

        let start_index = meta.to_linear_index(index.0.start, index.1.start);
        unsafe {
            let ptr = (self as *const T).add(start_index);
            &*<*const Mat<T>>::from_raw_parts(ptr as *const (), MatMeta {
                width: index.0.len(),
                height: index.1.len(),
                stride: meta.stride,
            })
        }
    }
}
```

> Note that we still need unsafety (via pointer arithmetic and dereferencing) in order to create a sub-matrix.

Reviewer: ```suggestion
Note that we still need unsafety (via pointer arithmetic and dereferencing) to create a sub-matrix.
```

#### Boxing a matrix

We could clone a `&Mat<T>` to a boxed `Box<Mat<T>>` by

> 1. Reduce the matrix into a slice

Reviewer: ```suggestion
1. Reducing the matrix to a slice
```

> 2. Box the slice

Reviewer: ```suggestion
2. Boxing the slice
```

> 3. Extracts the raw pointers, and replace the slice metadata by the matrix’s.

Reviewer: ```suggestion
3. Extracting the raw pointers, and replacing the slice metadata by the matrix’s.
```

> 4. Transmute the raw pointers back into a box.

Reviewer: ```suggestion
4. Transmuting the raw pointers back into a box.
```

    }
}
```

A bit-slice **cannot** be mutably sliced, however. This is because our architecture only support byte-level addressing.
If we mutate a bit, we need to copy the whole byte first.
So, even if two bit-slices are logically non-overlapping at bit-level,
their ends may still overlap at byte-level and would cause memory unsafety when mutated simultaneously.

```text
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|     [~~~~~~~~~|~~~~~~~~~~~)   |                   These two bit-slices are not overlapping at bit-level
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+                   but are overlapping at byte-level.

                +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                |           [~~~|~~~~~~)        |
                +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## Changes

This section provides instructions what needs to be added and changed in the standard library, compiler and language
to implement this RFC.

This RFC relies heavily on [RFC #2580].

### Core traits

We need the `core::marker::Pointee` trait from [RFC #2580], automatically implemented for all types:

```rust ,ignore
#[lang = "pointee"]

> #[fundamental]

Reviewer: What does fundamental on a trait mean (+ justify) (on all traits below that use it...).

pub trait Pointee {

>     type Metadata: Sized + Copy + Send + Sync + Ord + Hash + Unpin + 'static;

Reviewer: Justify these bounds somewhere.

Author: (Technically this should be justified in #2580; I'll just copy from the old text since it was there anyway 🙃)

```rust ,ignore
// Reduction on general custom DSTs are unsafe
pub unsafe fn reduce_ptr_unchecked<T: Derived + ?Sized>(val: *const T) -> *const Self::Reduced {
    let ptr = val as *const ();
    let meta = T::reduce_metadata(val);
    Self::from_raw_parts(ptr, meta)
}
pub unsafe fn reduce_mut_ptr_unchecked<T: Derived + ?Sized>(val: *mut T) -> *mut Self::Reduced { ... }
pub unsafe fn reduce_non_null_unchecked<T: Derived + ?Sized>(val: NonNull<T>) -> NonNull<Self::Reduced> { ... }
pub unsafe fn reduce_unchecked<T: Derived + ?Sized>(val: &T) -> &Self::Reduced { ... }
pub unsafe fn reduce_mut_unchecked<T: Derived + ?Sized>(val: &mut T) -> &mut Self::Reduced { ... }

// Reduction on regular DSTs are safe
pub fn reduce_ptr<T: Regular + ?Sized>(val: *const T) -> *const Self::Reduced {
    unsafe { reduce_ptr_unchecked(val) }
}
pub fn reduce_mut_ptr<T: Regular + ?Sized>(val: *mut T) -> *mut Self::Reduced { ... }
pub fn reduce_non_null<T: Regular + ?Sized>(val: NonNull<T>) -> NonNull<Self::Reduced> { ... }
pub fn reduce<T: Regular + ?Sized>(val: &T) -> &Self::Reduced { ... }
pub fn reduce_mut<T: Regular + ?Sized>(val: &mut T) -> &mut Self::Reduced { ... }

> ```

Reviewer: Which of these operations are derived and which are primitive?

Author: All of these are free functions provided in the standard library for convenience.

Reviewer: Ah; so they are all derived operations then?

### Syntax

Introduce a new kind of item.

```rust ,ignore
ItemKind =
    ... |
    CustomDst:{

>         "dyn" "type"

Reviewer: I think people will get the impression that these are type aliases somehow rather than new nominal types...
You might want to consider `dyn struct`.
Also, consider if you want `dyn(inline) type...` instead and if these could become attributes, e.g.:

```rust
#[repr(C), inline] // name already exists as an attribute so perhaps some other name...
pub dyn struct CStr(u8);
```

(generally if you can please the "no new syntax" people...)

Author: The problem (advantage?) of `dyn struct` is it makes it very natural to introduce

```rust
#[repr(C)]
#[dyn(inline)]
pub dyn struct Thin<T> {
    meta: T::Metadata,
    content: T,
}
```

however the type `Thin<T>::Reduced` will point to an anonymous struct 🤔

Reviewer: I'm not sure; I'm not nearly an expert on the contents of this RFC ;)

Perhaps?:

```rust
#[repr(C), dyn(inline)]
pub dyn struct CStr = OtherType;
```

Author: Lemme just shove all of these into the syntax bikeshed 😂

Reviewer: Sounds 🎵 good 🎸

When creating a *regular DST*, the compiler will automatically generate:

```rust ,ignore
dyn type RegDst<T, U>(regular: RegMeta<T, U>)
where
    T: Bounds<U>,
= ReducedTy<T, U>;

// implies:

impl<T, U> core::marker::Pointee for RegDst<T, U>
where
    T: Bounds<U>,
{
    type Metadata = RegMeta<T, U>;
}

unsafe impl<T, U> core::dst::Derived for RegDst<T, U>
where
    T: Bounds<U>,
{
    type Reduced = ReducedTy<T, U>;
    unsafe fn reduce_metadata(val: *const Self) -> Self::Reduced::Metadata {
        let meta = core::ptr::metadata(val);
        <Self as core::dst::Regular>::reduce_with_metadata(meta)
    }
}
```

When creating an *inline DST*, the compiler will automatically generate:

```rust ,ignore
dyn type InlineDst<T, U>(inline)
where
    T: Bounds<U>,
= ReducedTy<T, U>;

// implies:

impl<T, U> core::marker::Pointee for InlineDst<T, U>
where
    T: Bounds<U>,
{
    type Metadata = ();
}

unsafe impl<T, U> core::dst::Derived for InlineDst<T, U>
where
    T: Bounds<U>,
{
    type Reduced = ReducedTy<T, U>;
    unsafe fn reduce_metadata(val: *const Self) -> Self::Reduced::Metadata {
        <Self as core::dst::Inline>::reduce_with_ptr(val as *const ())
    }
}
```

> Users are expected to `core::dst::Regular` and `core::dst::Inline` themselves.

Reviewer: ```suggestion
Users are expected to implement `core::dst::Regular` and `core::dst::Inline` themselves.
```

Failing to do so should result in the normal [E0277] *“the trait bound is not satisfied”* error at the declaration site.

```
error[E0277]: the trait bound `CStr: std::dst::Inline` is not satisfied
  --> src/lib.rs:11:9
   |
11 | pub dyn type CStr(inline) = [u8];
   |                   ^^^^^^ the trait `std::dst::Inline` is not implemented for `CStr`
```

[E0277]: https://doc.rust-lang.org/error-index.html#E0277

The variance of the generic parameters `T`, `U` will be the same as the tuple `(Self::Reduced, Self::Metadata)`.

If `Self::Reduced` implements an auto-trait (e.g. `Send`/`Sync`), the custom DST itself will also implement it
(`Self::Metadata` is not considered).

The `#[repr]` attributes can be applied on custom DSTs.

* With `#[repr(align(n))]`, the `align_of` and `align_of_val` methods will always return `n`,
    and `Aligned` is always implemented regardless of `Self::Reduced`.
* With `#[repr(C)]` on an inline DST, the type is FFI-safe.
    `#[repr(C)]` on regular DST will cause error [E0517] *“attribute should be applied to struct, enum or union”*.

> * Everything else (packed, simd, transparent) causes [E0517] error.

Reviewer: I'm surprised by `transparent`; could you go into why?

Author: The `#[repr(transparent)]` is no-op. Added an option that allows it to be ignored.

[E0517]: https://doc.rust-lang.org/error-index.html#E0517

> Supporting `#[derive]` on custom DSTs is not required by this RFC.

Reviewer: But do you have a plan perhaps?

### Size and alignment

The bounds of the `align_of` intrinsic is relaxed to

```rust ,ignore
pub fn align_of<T: Aligned + ?Sized>() -> usize;
//                 ^~~~~~~~~~~~~~~~ relaxed
```

| Type | align_of |
|------|-------|
| `#[repr(align(n))]` | `n` |
| `#[repr(packed)]` | 1 |
| Sized type | (same as before) |
| Slices `[T]` | `align_of::<T>()` |
| `str` | 1 |
| Trait object | *not implemented* |
| Extern type | *not implemented* |
| DST structs/tuples | max of all fields, if all of them are aligned |
| Custom DST | `align_of::<Self::Reduced>()`, if the reduced type is aligned |

> Ensure `size_of::<&T>()` returns the size of the tuple `(usize, T::Metadata)`.

Reviewer: ```suggestion
Ensure that `size_of::<&T>()` returns the size of the tuple `(usize, T::Metadata)`.
```

The `size_of_val` and `align_of_val` of a custom DST is equal to its reduction.

| Type | size_of_val |
|------|-------------|
| Sized type | `size_of::<T>()` |
| Slices `[T]` | `val.len() * size_of::<T>()` |
| `str` | `val.len()` |
| Trait object | `metadata(val).size()` |
| Extern type | abort |
| DST structs/tuples | (some complicated calculations) |
| Custom DST `dyn type` | `size_of_val(reduce(val))` |

| Type | align_of_val |
|------|--------------|
| Aligned type | `align_of::<T>()` |
| Trait object | `metadata(val).align()` |
| Extern type | abort |
| DST structs/tuples | max of all fields |
| Custom DST | `align_of_val(reduce(val))` |

### Drop

> Relax bounds for the `needs_drop` intrinsic to check DSTs.

Reviewer: ```suggestion
Relax the bounds for the `needs_drop` intrinsic to check DSTs.
```

| Extern type    | false                       |
| Custom DST     | Same as `needs_drop::<T::Reduced>` |

The `drop_in_place` for custom DST will work by dropping its reduction in place.

```rust ,ignore
pub unsafe fn drop_in_place_for_derived<T: Derived + ?Sized>(val: *mut T) {
    drop_in_place(reduce_ptr_mut(val));
}
```

A custom DST itself cannot implement `Drop`.
Custom drop implementation should be done through its reduction, to prevent repeatedly dropping the same fields.

### Unsize

Modify `core::marker::Unsize` trait to:

```rust ,ignore
#[unstable(feature = "unsize", issue = "27732")]
#[lang = "unsize"]
pub unsafe trait Unsize<T: ?Sized> {
    const fn unsize(meta: Self::Metadata) -> T::Metadata; // <-- new
}
```

Error [E0328] *"explicit impls for the `Unsize` trait are not permitted"* should be removed.
Implementing `Unsize` is now allowed.

[E0328]: https://doc.rust-lang.org/error-index.html#E0328

Unsize-coercion from `*const Src` to `*const Dest` would perform the following:

```rust ,ignore
let src_meta = metadata(src);
let dest_meta = <Src as Unsize<Dest>>::unsize(src_meta);
unsafe { <*const Dest>::from_raw_parts(src as *const (), dest_meta) }
```

> Ensure the compiler can handle unsize-coercion between two DSTs, not just from sized to unsized.

Reviewer: ```suggestion
Ensure that the compiler can handle unsize-coercion between two DSTs, not just from sized to unsized.
```

//            ^~    ^~~~~~~~
//             |     underlying representation
//             |
//             metadata

dyn type Mat<T>(MatMeta) = [T];
//              ^~~~~~~    ^~~
//                    |     underlying representation
//                    |
//                    metadata
```

“`dyn`” is picked because it means dynamic,
and recently becomes a keyword thanks to `dyn Trait` ([RFC #2113]),
so we don’t need to introduce another unconstrained contextual keyword.
Adding a `type` makes it spell out “dynamic type”, as well as preventing potential misinterpretation as a `dyn Trait`.

### Customizing size and alignment

In the most flexible sense, the size and alignment of custom DSTs are arbitrary functions and will be provided by user.

```rust ,ignore
// not part of this RFC
trait CustomDst {
    fn size_of_val(&self) -> usize;
    fn align_of_val(&self) -> usize;
}
```

Now, while `std::mem::{size_of_val, align_of_val}` takes the whole DST reference as input,
it may not make sense for `align_of_val` to use its pointer as input. To wit, consider an unsized struct:

```rust ,ignore
struct S<T: ?Sized> {
    a: u8,
    b: T,
}
let s: &S<Dst> = ...;
let t: &Dst = &s.b;
```

> In order to get the address `&s.b`, we first need to find out the offset of `b` in `S<Dst>`.

Reviewer: ```suggestion
To get the address `&s.b`, we first need to find out the offset of `b` in `S<Dst>`.
```

Hence, our `align_of_val` should not depend on the pointer part of the reference at all.

```rust ,ignore
// not part of this RFC
trait CustomDst {
    fn size_of_val(val: *const Self) -> usize;
    fn align_of_meta(meta: Self::Metadata) -> usize;
}
```

The alignment needs to read `Metadata` because of how trait objects are implemented.
Consider the types `*S<u8>` and `*S<u16>`.

```text
+=======+    +===============+
| S<u8> |    |    S<u16>     |
+=======+    +===============+
| a | b |    | a |   |   b   |
+---+---+    +---+   +-------+
0   1   2    0   1   2   3   4
```

Both types can be *coerced* into `*S<dyn Debug>`.
The type `S<dyn Debug>` itself does not tell us the alignment of field `b`.
This can only be obtained via the type information stored in the trait object’s metadata.

### Reduction and regular/inline DST

We want to be able to treat every aspect of the a DST as if the ordinary DST to maintain safety.
So what the user should provide is not `size_of_val`/`align_of_meta`,
but a function which produces the original DST which we compute the size and alignment from it instead.

> We call such process *reduction*, and the custom DST is *derived* from that ordinary DST.

Reviewer: Should probably disambiguate "reduction" with "DST-reduction" or something since reduction is a more general concept (evaluation).

    type Reduced: ?Sized;
    unsafe fn reduced_metadata(val: *const Self) -> Self::Reduced::Metadata;
}
```

When we create a `dyn type`, the compiler should implement `Pointee` and `Derived` automatically,
and user just need to provide

```rust ,ignore
unsafe impl<T> Derived for Mat<T> {
    unsafe fn reduced_metadata(val: *const Self) -> usize { // the metadata of [T] is a usize, the length.
        ptr::metadata(val).len()
    }
}
```

The problem here is that we cannot forward `align_of_meta` to `reduced_meta`
since we don’t have `self` in the first place. This forces us to either

1. make `reduced_meta` take only `Self::Metadata`, or
2. make `align_of_meta` not rely on reduction i.e. the alignment is known at compile time.

We call the first kind of DST a *regular DST*. This covers slice, trait object, matrix, `OsStr`, and bit slice.
These types carry a nonzero metadata, and the reduction (and size) can be entirely derived from the metadata alone.

```rust ,ignore
trait Regular: Derived {
    fn reduce_with_metadata(meta: Self::Metadata) -> Self::Reduced::Metadata;
}
```

The second kind of DST are mainly thin DSTs like `CStr`, Pascal strings and length-prefixed arrays.

> These types has no metadata and thus the pointer is thin and can be used in FFI.

Reviewer: ```suggestion
These types have no metadata and thus the pointer is thin and can be used in FFI.
```

This also has a nice side effect of relaxing the bounds of `core::mem::align_of`.

Since we want to distinguish between regular and inline DSTs, and modify the syntax for declaring a custom inline DST.

```rust
dyn type Mat<T>(regular: MatMeta) = [T];
//              ^~~~~~~~

dyn type CStr(inline) = [c_char];
//            ^~~~~~
```

### Aliasing

Representing custom DSTs through reduction allows the compiler to manipulate them as if normal Rust types,
but there is a potential hazard regarding borrow-checking.

Let’s revisit matrix. A matrix’s metadata consists of three numbers, the width, height and stride.
The stride can be larger than the width when the rectangular slice does not include all columns.

```
+----+----+====+====+----+  highlighted region:
|  0 |  1 [  2 |  3 ]  4 |      width = 2
+----+----+====+====+----+      height = 2
|  5 |  6 [  7 |  8 ]  9 |      stride = 5
+----+----+====+====+----+
| 10 | 11 | 12 | 13 | 14 |
+----+----+----+----+----+
```

In the linear representation as a `[T]`, it will contain `[2, 3, 4, 5, 6, 7, 8]`.
Note that the irrelevant entries `4, 5, 6` are included.
This means providing mutable access to the reduced type is unsafe.

> Ideally we should represent this memory as `∃w,h,s: [([T; w], [Opaque<T>; s-w]); h]`,

Reviewer: Is this a sigma type? not super clear what the notation entails.
I think this means: `dyn<w: usize, h: usize, s: usize> [([T; w], [Opaque<T>; s - w]); h]`.

but encoding this thing in the type system is just as error-prone as writing actual checking code,

> and thus decided against doing this.

Reviewer: ```suggestion
and thus we decided against doing this.
```

This also means even if we have two `&mut Mat<T>`, their memory may be physically overlapping,
although they are still logically disjoint.
It is hard to know what LLVM is going to do with two `noalias` pointers pointing at different but overlapping memory,
if it turns out to cause mis-compilation ([issue #31681]), we may need to remove `noalias` annotations for custom DSTs.

[issue #31681]: https://github.com/rust-lang/rust/issues/31681

### Unsizing

Unsizing is a kind of coercion between two smart pointers `*T` and `*U`

> where both interprets the memory content in the same way, and `*U` is more general than `*T`.

Reviewer: ```suggestion
where both interpret the memory content in the same way, and `*U` is more general than `*T`.
```

```rust ,ignore
impl<T, const n: usize> Unsize<[T]> for [T; n] {
    const UNSIZED_METADATA: usize = n;
}
```

In [RFC #401] where unsizing was first introduced, unsizing is implemented for:

* Arrays to slices (`[T; n]` → `[T]`)
* Concrete types to dynamic trait objects (`T` → `dyn Trait`)
* Sized struct to unsized struct (`(u8, T)` → `(u8, U)` where `T: Unsize<U>`)

With custom DST, we wish to allow custom unsizing as well,
for instance it should make sense to coerce any `[[T; m]; n]` to a `Mat<T>`.

```rust ,ignore
unsafe impl<T, const width: usize, const height: usize> Unsize<Mat<T>> for [[T; width]; height] {
    const UNSIZED_METADATA: MatMeta = MatMeta {
        width,
        height,
        stride: width,
    };
}
```

Furthermore, it also makes sense to “unsize” an already unsized type, `[[T; m]]` → `Mat<T>`. And
thus the custom metadata should be an associated function, not an associated constant.

```rust ,ignore
unsafe impl<T, const width: usize> Unsize<Mat<T>> for [[T; width]] {
    const fn unsize(height: Self::Metadata) -> MatMeta {
        MatMeta {
            width,
            height,
            stride: width,
        }
    }
}
```

> Note that most unsizing implementations requires const generics ([RFC #2000]) to make sense.

Reviewer: ```suggestion
Note that most unsizing implementations require const generics ([RFC #2000]) to make sense.
```

}

struct ConstPtrFamily;
impl PointerFamily for ConstPtrFamily {
    trait MetaBounds = Copy + Ord + Hash + UnwindSafe + 'static;
    type Pointer<T: ?Sized> = *const T;
}

struct RefFamily<'a>(PhantomData<&'a ()>);
impl<'a> PointerFamily for RefFamily<'a> {
    trait MetaBounds = Copy + Send + Sync + UnwindSafe + Unpin + 'a;
    type Pointer<T: ?Sized> = &'a T;
}

struct RcFamily;
impl PointerFamily for RcFamily {
    trait MetaBounds = Clone + UnwindSafe + Unpin + 'static;
    type Pointer<T: ?Sized> = Rc<T>;
}
```

```rust ,ignore
trait Pointee {
    type Metadata<P: PointerFamily>: Sized + P::MetaBounds;
}
```

and a `Box<Dst>` will be represented as `(Box<()>, Dst::Metadata<BoxFamily>)` in memory.
This allows us to write the high-dimensional tensor slice as

```rust ,ignore
impl Pointee for Tensor<T> {
    type Metadata<P: PointerFamily> = P::Pointer<[(usize, usize)]>;
}
```

Since we can’t implement `Pointee` directly, we need to introduce a third kind of DST syntax e.g.

```rust ,ignore
dyn type Tensor<T>(third_kind<P>: P<[(usize, usize)]>) = [T];
```

The problem of generalizing `Metadata` is that it will plague every API it touches, e.g.

```rust

> fn metadata<T: ?Sized, P: PointerFamily>(dst: P

Reviewer: ```suggestion
to implement `Aligned` for `dyn X`.
```

Author: Suggested on the wrong line?

Reviewer: I think there's a bug or something in GitHubs UI because that's not the line I left the suggestion on...

Reviewer: Search for "in order"; I think this was one of these cases.

> fn metadata<T: ?Sized, P: PointerFamily>(dst: P

Reviewer: ```rust
trait T<U: Align<2>> {
    type V: PartialEq<U> + Align<2>,
}
```

;)

Author: 🤔 having `Align<2>` could imply a type can simultaneously implement `Align<1> + Align<2> + Align<4> + ...`

Reviewer: Oooooops!

```rust
trait T<U: Align<Val = 2>> {
    type V: PartialEq<U> + Align<Val = 2>,
}
```

We already have associated consts so we can probably do it today.
